### DISCUSSION TITLE
Delete Custom Instances Pretty Slow

### DISCUSSION THREAD

**Discussion by jmdelahanty** [talmolab/sleap#1891] [Help!]
Hey SLEAPy heads!

I thought my videos were solid with predictions but found a couple where my fish ended up in a spot I didn't expect/label very well, but still valid. I'm trying to use Labels -> Custom Instance Delete -> Delete Predicted Instances in Current Video

To delete about 57k predictions, it took around 5 minutes. I read a little about this kind of thing [here](https://github.com/talmolab/sleap/discussions/1058#discussioncomment-4219947) but wanted to see if that's expected. It's not a big deal that it takes a little bit to do, but just wanted to ask about it.

**Comment by roomrys** [comment]
Hi @jmdelahanty,

I started a little investigation and have not finished yet, but need to switch tasks now - will update when I get a chance. 

### Profiling
If you  run `sleap-label --profiling` then redo the action, we should be able to narrow it down to the function that is obviously taking the most time 
```
(sleap_1.4) me:~$sleap-label $dsdmc --profiling

Happy SLEAPing! :)
Saving config: /Users/me/.sleap/1.4.1a2/preferences.yaml
         223035 function calls (222552 primitive calls) in 32.377 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
      153    0.000    0.000    0.000    0.000 <attrs generated init sleap.gui.widgets.slider.SliderMark>:1(__init__)
        2    0.000    0.000    0.000    0.000 <attrs generated init sleap.instance.LabeledFrame>:1(__init__)
        1    0.000    0.000   32.377   32.377 <string>:1(<module>)
        1    0.000    0.000    0.000    0.000 QtGui.py:35(__getattr__)
        1    0.000    0.000    0.002    0.002 __init__.py:215(dump_all)
        1    0.000    0.000    0.002    0.002 __init__.py:248(dump)
        1    0.000    0.000    0.000    0.000 _bootlocale.py:33(getpreferredencoding)
```
My readily-available test dataset isn't so big so all the times are pretty small, but it would be more prevelant in your large dataset.

### Code Investigation
Immediately after click "Delete Predicted Instances in Current Video", we loop over all `LabeledFrame`s and all `Instance`s in each labeled frame:
https://github.com/talmolab/sleap/blob/d3ad2261990a30effdbe568290c772fd3f6d5cc9/sleap/gui/commands.py#L2324-L2334

. Then, we confirm the deletion with the user and start the actual deletion where we loop over only the `LabeledFrames` and `PredictedInstance`s to delete:
https://github.com/talmolab/sleap/blob/d3ad2261990a30effdbe568290c772fd3f6d5cc9/sleap/gui/commands.py#L2294-L2298

during which we call `Labels.remove_instance`
https://github.com/talmolab/sleap/blob/d3ad2261990a30effdbe568290c772fd3f6d5cc9/sleap/io/dataset.py#L1318-L1322

which calls the [`List.remove` method](https://github.com/python/cpython/blob/fda6bd842a2b93a501526f1b830eb900d935ac73/Objects/listobject.c#L3257-L3278) (another iteration through only `Instance`s in that particular `LabeledFrame` until it finds the `Instance` to remove - at worst iterating through all the `Instance`s in the `LabeledFrame.instances` list).

At this point, the `PredictedInstance`s have been deleted. While removing `PredictedInstance`s, we were also keeping track of now-empty `LabeledFrame`s to remove which is what we will do now
https://github.com/talmolab/sleap/blob/d3ad2261990a30effdbe568290c772fd3f6d5cc9/sleap/gui/commands.py#L2301-L2304
.

Incomplete investigation, but will return with an update.

Thanks,
Liezl
